Add nanosecond resolution support on unixes to `jl_gettimeofday` and `time()`

Seelengrab opened the pull request:
Windows apparently only has millisecond resolution guaranteed, per their API (https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/ftime-s-ftime32-s-ftime64-s).

Fixes #44890.

Seelengrab committed:
Add nanosecond resolution support on unixes to `jl_gettimeofday` and `time()`.

Windows apparently only has millisecond resolution guaranteed, per their API (https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/ftime-s-ftime32-s-ftime64-s).

Seelengrab commented on src/support/timefuncs.c:
```diff
@@ -37,7 +38,7 @@ JL_DLLEXPORT double jl_clock_now(void)
+    return now.sec + now.nsec * 1e-9;
 }
 
 void sleep_ms(int ms)
```
Should the `select` in this function really be used here? I think using a different API for sleeping like `nanosleep` would be more appropriate, as it can also consume seconds & nanoseconds like this PR enables via `time()`.

Seelengrab commented on src/support/timefuncs.c:
```diff
@@ -7,7 +7,7 @@
 #include <sys/timeb.h>
 #else
-#include <sys/time.h>
+#include <time.h>
```
Not 100% sure about this, but was required to get it to build on my system. CI will tell if this was a mistake :shrug:

Keno commented on base/libc.jl:
```diff
@@ -137,7 +137,7 @@ systemsleep
 
 struct TimeVal
```
Do we consider this breaking? If so, we may want a getproperty overload that provides access to the old field name.

Seelengrab replied:
I don't think so - it's neither exported, nor does it show up on docs.julialang.org (as far as I can see).

Keno replied:
It's not exported from Base, but `Libc` is and people are using these as `Libc.TimeVal`, e.g.:
https://github.com/ancapdev/LightBSON.jl/blob/729b4308793b26dd4fa6635c224b8165299ff6a4/src/object_id.jl#L69
https://github.com/ancapdev/UnixTimes.jl/blob/master/src/UnixTimes.jl#L104
https://github.com/timholy/Revise.jl/blob/master/src/utils.jl#L111-L115

Seelengrab replied:
Right, I did change the type from `timeval` to `timespec` - I should leave that struct alone and add `TimeSpec` to `Libc` instead, with a conversion from `TimeSpec` (and an deprecation notice for `TimeVal`?)

How did you come up with that list? I think all of these should have used `time()` in the first place, rather than manually calling into `Libc` on their own :thinking: I may make PRs fixing that, if the list isn't too long..

Seelengrab replied:
Trying that deprecation, I don't think I can easily add that do `deprecate.jl` using `@deprecate` since that defines a new method, which I don't necessarily want if I want to "hook into" the new `TimeSpec` and just convert it - how should I go about that?

Seelengrab replied:
Well, I can't deprecate the constructor, since `depwarn` only takes a symbol, not a signature...

Seelengrab replied:
Ok, UnixTimes and LightBSON can be patched/worked around once this PR hits, Revise has a PR open now and I'll open an issue for InterProcessCommunication.jl once this hits. The struct has to remain for now, but at least one constructor is marked deprecated. I'll add the deprecation notice of that to `NEWS.md` as well.

PallHaraldsson commented:
Maybe we want to rather use the faster CLOCK_REALTIME_COARSE? Or if actually want nanosecond resolution, do that on Windows too (as seems possible). I'll explain on the open issue.

Seelengrab commented:
> do that on Windows too (as seems possible)

How so? As far as I know, there's no other API that's convenient to use, no?

Seelengrab committed:
Add `TimeSpec` and deprecation notice for `TimeVal`.

`TimeVal` is used by some directly, to then call back into libc manually instead of using `time()` for cross platform compatibility. This preserves the original fields by adding a conversion from `TimeSpec` and a deprecation notice of `TimeVal` instead.

Seelengrab commented:
```
Error in testset read:
Error During Test at none:1
  Got exception outside of a @test
  ProcessExitedException(2)
  Stacktrace:
    [1] try_yieldto(undo::typeof(Base.ensure_rescheduled))
      @ Base ./task.jl:877
    [2] wait()
      @ Base ./task.jl:941
    [3] wait(c::Base.GenericCondition{ReentrantLock})
      @ Base ./condition.jl:124
    [4] take_buffered(c::Channel{Any})
      @ Base ./channels.jl:416
    [5] take!(c::Channel{Any})
      @ Base ./channels.jl:410
    [6] take!(::Distributed.RemoteValue)
```

should be unrelated?

KristofferC commented:
Looks fairly straightforward to me. Needs a rebase though.

Seelengrab committed:
Add deprecation notice of `Libc.TimeVal` to `NEWS.md`

Seelengrab commented:
Having to rebase due to one of my own PRs creating a conflict in NEWS.md - not sure if I should be happy about it or not :joy:

vtjnash committed:
Merge branch 'master' into jl_gettimeofday

vtjnash commented on base/libc.jl:
```diff
@@ -176,13 +176,20 @@ struct TimeVal
+
+function TimeSpec()
+    ts = Ref{TimeSpec}()
+    status = ccall(:jl_gettimeofday, Cint, (Ref{TimeSpec},), ts)
```
This could now go through libuv to avoid our need for a custom wrapper (and get 100-nanosecond precision)
```suggestion
    UV_CLOCK_REALTIME = 1
    status = ccall(:uv_clock_gettime, Cint, (Cint, Ref{TimeSpec},), UV_CLOCK_REALTIME, ts)
```
https://docs.libuv.org/en/v1.x/misc.html#c.uv_clock_gettime

Seelengrab replied:
Does this have resolution guarantees? I can't seem to find that in the libuv docs :thinking:

vtjnash commented on base/libc.jl:
```diff
@@ -176,13 +176,20 @@ struct TimeVal
-    status = ccall(:jl_gettimeofday, Cint, (Ref{TimeVal},), tv)
+struct TimeSpec
+   sec::Int64
+   nsec::Int64
```
```suggestion
   nsec::Int32
```

This is actually a Clong, which is sometimes Int32 and sometimes Int64, and sometimes just plain inconvenient to know what the platform will do. But if we use libuv, then it is explicitly Int32.
